### DISCUSSION TITLE
How to do cross-field validation?

### DISCUSSION THREAD

**Discussion by croraf** [react-hook-form/react-hook-form#7152] [Q&A] [answered]
Hi folks. I have problems with cross-field validation.

I have a checkbox and a textarea. If checkbox is checked then filling textarea is optional, but if checkbox is unchecked filling textarea is required. How can I do that?

I've tried something like this:
```ts
<TextField
        {...register('freeText', {
          validate: (v) => watch('myCheckbox') === true || !isEmpty(v)
        })}
        multiline
        rows={10}
        style={{ width: '100%', marginTop: '2rem' }}
        label={'Text'}
        InputLabelProps={{ shrink: true }}
      />
```
But it doesn't work. When `myCheckbox` gets checked this isn't updated and it still shows an error.

**Comment by fahadsohail482** [comment, marked as answer]
Hi @croraf,
This can be done easily by providing a validation schema to useForm hook.
The Clean approach would be to simply add the condition to your text area if the checkbox is false text area is required.

Here is the [working CodeSandBox](https://codesandbox.io/s/practical-grothendieck-zcpme?file=/src/App.js) Implementation of your problem.
I hope this solves your Problem.

![image](https://user-images.githubusercontent.com/46647496/143525987-defff11b-1de4-46a1-af53-791aaf8b8b9c.png)

**Reply by bluebill1049** [reply]
thank you @fahadsohail482 for sharing solution

**Reply by croraf** [reply]
Thanks @fahadsohail482 
There is one problem here.

Check this test:
1. Click submit. It will not let you submit the data and will show an error.
2. Now check the checkbox. This should remove the error. But it does not!
(I can submit it though and the error disappears when I start typing into the text field)

**Reply by getTobiasNielsen** [reply]
I think you need to make separate check on state, because you want to know what state you are currently in.

Let me know what you guys think!

https://codesandbox.io/s/llmxe?file=/src/index.js

**Reply by fahadsohail482** [reply]
> Thanks @fahadsohail482 There is one problem here.
> 
> Check this test:
> 
> 1. Click submit. It will not let you submit the data and will show an error.
> 2. Now check the checkbox. This should remove the error. But it does not!
>    (I can submit it though and the error disappears when I start typing into the text field)

@croraf let me have a look.

**Reply by fahadsohail482** [reply]
@croraf Please have a look at the updated version. [Updated Version](https://codesandbox.io/s/practical-grothendieck-zcpme?file=/src/App.js)
I have to use Controller for the checkbox and when we change the checkbox it triggers the validation.
For the old version, it seems to have some sort of revalidation bug for the checkbox.
Hoping that this will help you solve your problem.
![image](https://user-images.githubusercontent.com/46647496/143619930-9ae45d88-28b7-446c-bd04-46fd8b040745.png)

**Reply by getTobiasNielsen** [reply]
@fahadsohail482 I updated my example, you may want to see the approach. I think you want to trigger only when you make it optional.

**Reply by fahadsohail482** [reply]
@getTobiasNielsen the requirement here is to make the text field optional when the checkbox is true and make it required when the checkbox is false.
I want to trigger every time when the checkbox is checked or unchecked to revalidate and update the error object.
![image](https://user-images.githubusercontent.com/46647496/143671586-68e3adbd-6068-40d5-b43a-181b9d8fe7a6.png)


I have updated your example [here](https://codesandbox.io/s/react-hook-form-watch-v7-forked-cex2o?file=/src/index.js)
